XiEditor/Storage/SettingsProvider.swift

```diff
@@ -0,0 +1,36 @@
+//
+//  SettingsStore.swift
+//  XiEditor
+//
+//  Created by Alex Demishkevych on 10/13/18.
+//  Copyright © 2018 Raph Levien. All rights reserved.
+//
+
+import Foundation
+
+protocol SettingsProviding {
+    func fetchingRecentlyUsedLanguages() -> [String]
+    func saveRecentlyUsed(languages: [String])
+}
+
+class SettingsProvider: SettingsProviding {
```

Comment:
This architecture, it seems needlessly convoluted. Yes, it looks a bit like MVC but that falls apart because in practice you can only access the `SettingsProvider` through the `SettingsProviderProcessor`. I'm not sure what benefits this separation is supposed to bring.

Also having `SettingsProvider` and `SettingsProviding` which are almost the same thing is confusing. I'd rather call the protocol `SettingsProvider` or even `ISettingsProvider` (to clearly indicate that it's an interface/protocol) and this class something like `SettingsStore`, since all it does is provide access to the KeyValueStore. Or you could probably remove this class entirely since it's just a very thin wrapper over a dictionary.

Reply:
@jansol to be honest I did not get what you meant by `MVC` and access the `SettingsProvider` through the `SettingsProviderProcessor`. Could you please explain what did you mean?

The idea behind `SettingsProvider` is that it responsible for only for interacting with storage without any additional logic while `SettingsProvidingProcessor` decorates `SettingsProvider` adding level of logic.

This solutions are flexible from my point of view, because `SettingsProvider` implementation can be easily replaced with new one (for example new implementation could fetch settings from `SQLite` instead of `UserDefaults`) and injected into `SettingsProvidingProcessor`. So, final configuration has new provider with new storage and with same processor that contains unchanged languages processing logic. As the project evolves, `UserDefaults` may not be enough and switching to some type of data base could be done quite easy.

What do you think? Does it worth to have such solution?

Yes, naming definitely should be improved. I would not use `ISettingsProvide` since it is not a best practice for naming protocols in Swift. Also I am trying to use this guideline during API design [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/).